### DISCUSSION TITLE
Open-Graphics: Implementation.

### DISCUSSION THREAD

**Discussion by Sheltur** [OpenRCT2/OpenRCT2#15892] [Feature requests]
This is technically more a discussion on the best methods for implementing Open-Graphics.

Obviously we have several routes we can take.

1. Route one is quite simple, Just a basic replacement of all graphics that can be toggled on/off. This method is clean, simple and very basic, yet it does have its downsides. The major upside is how simple it would be to implement. The issues are that it is far the most limiting of implementation methods. As RCT is half creative focused sandbox, half business simulator, it really limits the creative aspect of the game by not allowing players to use both assets from open graphics and the original in conjunction with one another. 
2. Dynamic replacement: Objects get replaced on a client-side level depending on if a user has the assets installed or not. This allows for both Open Graphics Objects and Original Graphics Objects, to be placed aside one another, giving players far greater creative expression and choice. While also allowing those who do not have the original graphics sets installed to still see and play with other users that do. This method also allows for recoloring of once non-recolorable objects, as these new open graphics alternatives during their swap will be forced to a color that matches closest to the original asset and cannot be recolored.
3. Hybrid Approach: As some assets cannot be fully replaced, Guests, Water, UI And various other non-selectable or interchangeable objects will still need to be full on replaced. And this can be toggled in the option's menu. This personally I believe is the best method for Open-Graphics Implementation, allowing for the most flexibility while still maintaining ease of use for players.

Now this is from my perspective, I have no idea how much effort a method such as this will take to implement. Which is why I am requesting it this early into the project while the project is still in its infancy. I believe a method like this will give the most creativity not just to the players but also the artists such as myself who will be making the assets. 
I'd love to hear any feedback on what you think about this as a whole.

**Comment by oli414** [comment]
What we've discussed with other devs is a system that is similar to that of OpenTTD. Where base graphics come as a set that can be applied to the game. A map or server would likely specify a preference for a certain set. Defaulting to OpenGraphics when the specified set is not available.
![image](https://user-images.githubusercontent.com/2348094/144470927-40c2e811-e0f5-4994-8de9-2d797ec05529.png)

The original graphics would be a graphics set that requires the original game to be linked in order to be used. Once OpenGraphics is finished it is possible that there's an influx of people who do not have the original game, which is important to keep in mind.
The OpenGraphics replacement set would thus become the default to remove the requirement for the original game.

Since it essentially acts as a "texture pack" of sorts, it wouldn't be possible to play the OpenGraphics and original variants of a specific scenery object next to each other. I think this would greatly complicate things if we did allow it.

> As some assets cannot be fully replaced, Guests, Water, UI And various other non-selectable or interchangeable objects will still need to be full on replaced.

I'm not sure what you mean by that, but OpenGraphics will replace G1 graphics as well, which includes guests, water, and UI.

OpenGraphics will once finished always be the most complete graphics set, containing replacements for the original objects as well as new things added by OpenRCT2. As such it would act as a fallback for any graphics set much like the English localization file is a fallback for all languages.

**Reply by Sheltur** [reply]
Ah, okay! I think you might have just misread, or maybe I didn't explain it in a clear enough way. Basically, I'm suggesting we don't do what you are suggesting, and instead we basically swap objects if they aren't installed. So we can still use open graphics objects alongside the original graphics objects. This gives us the creators the most flexibility and also the players the ability to use them alongside one another.

Basically, When a save/scenario/server loads, the game does a check to see if you have the original graphics installed, and preforms a visual swap on the objects. So if a player joins a server using both original graphics and open graphics objects, let's say they like the look of the original tree assets and a few coast tracks for instance, The player that doesn't have the OG assets installed will just see the open graphics. While those who have both installed will see both based on what is selected, basically it makes the game playable for those who don't have it while also giving the most flexibility for those who do.

Now! Not all objects can be swapped on the fly like this, which is why I mentioned assets such as guest water and the UI, These would be swapped the same way Open TTD does it in the menu, Basically we would call it. “UI Graphics” for instance and we can swap between them, Maybe a separate option for "Open-Guests" and maybe we can swap water the same way we swap color palettes? But I'm not 100% sure how we would do that or if that's the best option for water.

I think that this “on the fly replacement” method provides the most flexibility for players and also still provides the seamless experience you expect for those who don't have the OG graphics installed.

This is why I think a sort of hybrid approach to going about this is best, OpenTTD is a business management game first and foremost, RCT2 is more... 50% Creative sandbox, 50% business management sim. Which is why I think others would like the ability to select both.

So, Basically how you would select a current theme when setting up a scenario, you will have both options, Open graphics will basically be treated like “an expansion” In the sense that it adds its own new content. And themes for players to choose  from, Now these themes would represent themes that already exist and are designed to replace them yes, but doesn't strictly force a player to choose one over the other if that makes sense.

Now, maybe it would probably be best once open-graphics comes out to “flip it” in the sense that open-graphics is the default and the original game we treat as an expansion.

Anyway, I hope that sort of clears it up, that instead of treating it as a texture pack that we just drop in, we treat it like an expansion that can dynamically replace missing assets!

**Reply by oli414** [reply]
I understand what you're suggesting:

> Since it essentially acts as a "texture pack" of sorts, it wouldn't be possible to play the OpenGraphics and original variants of a specific scenery object next to each other. I think this would greatly complicate things if we did allow it.

I'm just trying to explain what the current plan looks like. As that would not be able to support what you're suggesting at the moment.

During the transition period of OpenGraphics, before all the replacement graphics the original graphics will remain the fallback. Ideally allowing the people working on OpenGraphics to load the replacement assets that are available to replace the original graphics where possible.

In the larger scheme of things, as I said, I personally don't think the OpenGraphics should be implemented as separate objects. It would greatly complicate the OpenGraphics project, as the replacement graphics will both act as a fallback for original objects, as well as new objects simultaneously. Additionally, that would imply that OpenGraphics is not the default, which I don't think will happen. I don't see much value in having both original and OpenGraphics objects simultaneously as the OpenGraphics is designed to be a replacement. So you end up with a lot of very similar but not identical items.

I'd argue that that would cause the project to remain too reliant on the original graphics, which defeats the entire purpose of OpenGraphics. As people would continue to use original graphics along side OpenGraphics, making it harder for new comers to view those hybrid parks.

Consider OpenTTD, most people do not own TTD to allow them to use the original graphics. I'd imagine that (and I know the community will tell my otherwise) over time the original graphics will phase out to the point where it is only used by legacy members of the community, and to open pre-OpenGraphics parks.

**Reply by Sheltur** [reply]
I honestly think having them as separate objects and just falling back to them when needed provides a nice amount of flexibility for players. Also, it gives us as artists more flexibility as well. I think over all, what we should be focusing on is not just making suitable replacements, but also things that compliment them well. So the problem with a one to one replacement, is it makes adding new objects to pre-existing themes pretty confusing. But if we made a sort theme that rhymes with the original themes and also compliments the originals. I do know and understand that as time goes on, the original assets are going to be less common, but I think still giving the players the option to have both is a nice addition.

I think as time goes on I can see players asking for something like this, and instead of what I think will end up happening, where people just make CSO that lets them add both at the same time. Not only that, but I think just supporting it right out the gate is a nice addition.

I also don't really agree that people would assume the original graphics are the default, and It's more just an option for those players that want it.

The biggest reason, I think you might be currently… “Assuming” that open graphics will be a very close almost one to one remake of the original assets, which they won't be. Mainly because we can't for obvious legal reasons. Open graphics will not and won't ever look identical to the originals. Will they fill the same role as the originals? Yes they will! Will they also look like they fit into their original themes? Yeah! They will as well! But we won't ever have an exact replacement for the originals. And letting players have the option to do so if they wish? I think that is something people will find extremely useful.

I do feel like a bit of a broken record, I know, and I know I can't really convince you over text. Maybe we could hop into voice chat at some point and I can try to get the other reasons across and maybe get a back and forth going as someone who plans on helping with Open Graphics.

Still, thanks so much for at least taking your time to respond to me and read what I have to say!

**Reply by oli414** [reply]
No worries! I get what you're saying, 

I don't expect Open Graphics to be an exact remake at all by the way. There will be huge differences and we'll have to accept that.

It'd be a good idea to get some more opinions on this. I'm personally not a fan, but perhaps if a lot of folks do agree then we can look into changing our plans.

**Comment by duncanspumpkin** [comment]
At present we also do fallbacks for rct1 graphics if that is not linked and an rct1 feature is used.

**Reply by Sheltur** [reply]
This is more what I am suggesting for open-graphics, if the original game's graphics are not linked we fall back to Open graphics. While also giving the players the options to select both if they want, just as how we can use both the RCT1 dirt paths* and RCT2 dirt paths at the same time if you really want to for what ever reason.

**Reply by ZehMatt** [reply]
I think the idea is to have OpenGraphics being the default and RCT1/RCT2 being optional not the other way around. This is of course not possible at the current state.

**Comment by oli414** [comment]
@IntelOrca curious to hear your thoughts on this

**Reply by IntelOrca** [reply]
The easiest way to implement this at the moment is the hybrid approach. The game would just load an OpenGFX object instead of the vanilla one for a given ID. But hybrid would be possible if you were to add the OpenGFX objects by their new ID directly.

**Reply by oli414** [reply]
I feel like there're some unwanted consequences that have to be considered. For example:

If a user opens a park made with original graphics in a game with no original graphics. The graphics will be OpenGFX. The user now expands a building using the same walls that were already there (original walls that have OpenGFX graphics replacements).
Another user that does have original graphics installed now opens the park and sees a building that is made out of 2 different wall textures.

From a user point of view, I can't imagine a way of presenting this cleanly. There are a lot of edge cases that wouldn't be the case with a non-hybrid approach.

**Reply by oli414** [reply]
Another example:

A server has both the OpenGFX as the original variant of a flat ride selected in the object selection.
A user joins who doesn't have the original graphics now either sees 2 of the same flat rides in the ride selection screen. Or is only able to select the OpenGFX variant.

The only way I could see this working is if original objects are locked away instead of being replaced by OpenGFX when the originals are not available. But that would divide the community.

@Sheltur do you have any thoughts on how this could be resolved?

**Reply by Sheltur** [reply]
Division of the community is a great criticism of this method, In the rides and objects menu I would probably assume it would just simply not show the rides that they don't have installed. So let's say they have two scrambled egg rides in their park. The users with Open Graphics only installed would receive a message box pop up when joining a server or save file that says something along the lines of. “Some objects in the park have been replaced” along with a link to the openRCT2 website going into further detail as to why. Alongside a close option and a checkmark with “Don't show me this message again.” option. And when ticked, the pop-up won't appear the next time you load in, similar to how The Sims has done it with Custom Content.

Now this isn't a perfect way of going about doing this, I know, but I think the sacrifice is worth it as opposed to just limiting users to who do have both to only picking one.

For Scenery objects, It would work pretty similar, The players without the OG-GFX installed will only be able to see and place the Open-GFX scenery objects. Basically, we would just hide the OG-GFX themes and objects for those users, as they wouldn't be able to see the difference anyway. Just how we currently do it for RCT1 graphics. If a park has RCT1 paths, maybe because they like the ramps more than the stairs they can still use them, but the players who just have RCT2 installed will see stairs for both.

I do get that having this disconnect does kind of suck? But I don't think It's as big of a deal as you are expecting, Hopefully if we make our assets right, players without them won't even notice what has and has not been replaced.

And yeah, as the above comment explains, the Open Graphics will be the default once completed and adding in RCT2's assets works like how we do it with RCT1's currently. 

Now, as also talked about, Some objects will just straight up need to be replaced out right. And that's fine. 

Division of the community: I think that this is just a necessary out-come of open graphics in general. Even if works like OpenTTD, I think you'd be surprised how many players might just keep using the OG-GFX. And that same disconnect will be there. People playing using their original graphics and a new player joins in and is confused why the other player doesn't like what they are placing down because it doesn't look how it does on their end. With the hybrid approach this will happen 50% less just because it's a one way street rather than a two-way one.

Now you don't have this issue as much with OpenTTD because its main focus is transportation and business management. But OpenRCT2 isn't mostly a theme park tycoon game, If anything most players in the current day play it as a co-operative sandbox. We don't even currently have an option of playing Vs. one another. (Not to say we shouldn't as that sounds cool af lol)

Anyway, that's my personal outlook on how we would go about doing this. It's not a perfect solution, but I do think it is the one with the least amount of downsides.

**Reply by oli414** [reply]
Yeah that could work!

> Hopefully if we make our assets right, players without them won't even notice what has and has not been replaced.

Doesn't that defeat the purpose of this proposal? What is the benefit of having a hybrid method if the players won't notice which is which?

Without the hybrid approach it'll be easier to give people an incentive to move away from OG-GFX. As new features or track pieces get added. The OG-GFX will come more and more at a disadvantage.

Regarding division though given the example I gave. A person joining without OG-GFX installed would have a better chance of building something along with the other players that do use OG-GFX. If the server is configured for OG-GFX in a non-hybrid approach. The player without OG-GFX would simply see everything as OpenGFX. This also means that they can place down walls that match with the walls that the user with OG-GFX already placed down.

Surely there are some differences in what both users see, but at least they can still build together. Additionally this approach has a higher chance of people ending up just using OpenGFX as users that are not familiar with the hybrid system won't accidentally select OG-GFX objects that give a bad experience for non OG-GFX players.

Just trying to go over what the end experience looks like for the users. Again I certainly see value in being able to have more objects essentially, but I'm afraid it'll come at the cost of making the user experience sub-optimal by heavily overcomplicating how people expect the game to work, especially in regards to how other people see the game and can play the game.

**Reply by Sheltur** [reply]
> Without the hybrid approach it'll be easier to give people an incentive to move away from OG-GFX. As new features or track pieces get added. The OG-GFX will come more and more at a disadvantage. 

Actually I'd say if anything, this is the other way around! Using the OpenTTD method for this approach means that swapping back to OG-GFX causes new track pieces to become missing, as these assets don't support the new elements. But having two different “Corkscrew” coasters, one that is “open-GFX” and the other “OG-GFX” means that we can add lets say sloped banked turns to the track. Something a lot of people want. The original graphics do not have those assets to choose from. So if we swap back, those now supported track pieces become invisible.

 But having both options available means that if you load an old save file. The corkscrew coaster looks and acts just as it did the day it came out, but our new corkscrew coaster has all the added pieces players want.

> Doesn't that defeat the purpose of this proposal? What is the benefit of having a hybrid method if the players won't notice which is which?

No I don't think so, as I said the pieces won't be one to one, but they will still be designed to fill the same role! So we will need, let's say… one of those quarter tile evergreen trees. It will be about the same height, the same size the same type of tree as the original, but it will still look different. Or let's go ahead and say we will have… Roof tiles in the generic theme, A lot of players want these to be recolorable, but the OG-GFX isn't. We can make the new OpenGFX roofs recolorable while having the originals not be. (How do we go about swapping these, you say?) We just force recolor them to match the original when swapped, and we won't let the OpenGFX only players recolor them. And when they try it gives one of those little red error boxes saying it can not be recolored. (Now I understand this is kind of a janky way about doing things? But I think the new added functionality is worth the headache.)

Or let's take an object that's already different, The pizza stall, we already have a new pizza stall object designed. But it looks nothing at all like the original, Maybe someone wants both? They both fill the same gameplay role, yes, but they do look different. Suddenly swapping into OpenGFX and seeing all their pizza stalls look different might be confusing for players who are migrating over their old save files and swap into OpenGFX.

Each object will be about the same size, about the same style, and the same in terms of price. But they will be still visually distinct from one another, even if they fill the same exact role.

And yeah, they would see everything as OpenGFX, But they will still match the originals. The idea is that they look like they belong aside them, like a sibling piece. A player building with OG-GFX in the Japanese theme, for instance, places down the OG pagoda. The OpenGFX pagoda still fits with it. Will the OpenGFX only players see both as the OpenGFX yeah, will the OG-GFX see them as two separate pieces, yep! But they also should try to match with one another. 
It just gives those who have the original game more options and doesn't force people to choose the new graphics with more features, over the old nostalgic graphics they love.

It's really about striking a balance between something that is visually its own while also filling the same role as the original.

Think of it like being partly colorblind. Are you technically missing something? Sure, but the point is would they notice if no one told them what things are supposed to look like? Probably not. They would just assume that this is how things look.

**Reply by X123M3-256** [reply]
> Actually I'd say if anything, this is the other way around! Using the OpenTTD method for this approach means that swapping back to OG-GFX causes new track pieces to become missing, as these assets don't support the new elements

This is the behaviour I want it to have. If the new sprites can be used alongside the originals that means they have to match almost perfectly, which is extremely difficult to do. Honestly, this is the main reason I'm interested in OpenGraphics in the first place - if it's not a binary choice between OpenGFX or vanilla, then it's of no use to me.

You may think that giving players more choice is better, but bear in mind that making it harder to create good sprites may well limit the amount of sprites that get made, and there aren't exactly a lot of people making sprites to start with. To me, the main attraction of the OpenGraphics project in the first place is the removal of this restriction. I could add sloped banked turns to the corkscrew coaster right now - I've automated the process to the point where I can create a complete set of track sprites in half an hour. But as long as the new sprites have to work alongside the originals, it will never be worth the effort for me.

**Comment by 733737** [comment]
currently the documentation says use blender 2.7 for pipeline, but blender 3 has recently released. Any thoughts on whether we should update documentation or remove version reference?

**Reply by Sheltur** [reply]
Myself and a few other people I have talked with would also like to use Blender 3.0 as the foundation for the pipeline. But several of the other people working on the project believe that 2.79s blender render has a more accurate approach with lighting and dithering on renders. Personally, I think we can achieve the same or similar effect in blender 3.0 with proper shader and lighting set-ups. Currently, this is up in the air, If you plan to make a model for OpenGFX feel free to use Blender 3.0 as we can always re-render in 2.79 later on if we don't like the looks of 3.0s renderer.

**Reply by oli414** [reply]
There is a lack of evidence that Blender Cycles or Eevee can be setup in a consistent manner that replicates the style well.
Blender Render (removed after 2.79) is the same tech that Simon Foster used for his renders (scanline renderer) and due to its simplicity creates more predictable results.

I would recommend sticking with 2.79 until there's more evidence that Eevee or Cycles can be used effectively. Benchmark tests for those renderers are currently lacking, apart from my own benchmarks that I have done, which I deemed to be unsatisfactory.

This post shows how severe the differences can be:
https://blenderartists.org/t/quality-of-rendered-sprites-2-8/1156814

Blender 2.79 files and later versions or vice versa are not interchangeable due to the large differences in the material setup.

To move the project forward, Blender 2.79 was decided as the version to use. As the project would otherwise be forever stuck in the version discussion which has been ongoing for years now. This can change in the future, but will require more research. Until we have consistent benchmarks to prove that Blender 3.0 can do the job we will stick with 2.79.

**Comment by ThiaJay** [comment]
'Has Simon Foster contributed towards the graphics used in OpenRCT 2? Is there a reason no one seems to have the original 3D renders?

**Reply by oli414** [reply]
No, Simon Foster made the graphics for the original game (RCT2). The graphics were pre-rendered. Meaning that only the renders (2D sprites) were shipped with the game. This is also why we do not have the original 3D models.

**Comment by ThiaJay** [comment]
I've seen several people hint at Simon still having the original 3D models and one of his videos on YouTube along with comments on reddit also indicate this.

**Reply by oli414** [reply]
I have spoken to Simon, he's likely not able to share them contractually, and besides that he's lost the original files.

Anyways, your question is not within the scope of this discussion. Feel free to open a new discussion if you want to discuss this further.

**Reply by ThiaJay** [reply]
Thanks for confirming Oli.
